File: internal/discovery/discovery.go
package discovery

import (
	"bufio"
	"bytes"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/bee/java-process-mapper/internal/model"
	"gopkg.in/yaml.v3"
)

type Options struct {
	RootPath     string
	JavaVersion  string
	IncludeTests bool
}

func Discover(opts Options) (*model.Project, error) {
	root, err := filepath.Abs(opts.RootPath)
	if err != nil {
		return nil, err
	}
	info, err := os.Stat(root)
	if err != nil {
		return nil, err
	}
	if !info.IsDir() {
		return nil, fmt.Errorf("root path is not a directory: %s", root)
	}

	project := &model.Project{
		Name: filepath.Base(root),
		Root: filepath.Clean(root),
	}

	modules, err := discoverModules(root)
	if err != nil {
		return nil, err
	}
	if len(modules) == 0 {
		modules = []model.Module{{
			ID:   stableID("module", root),
			Name: filepath.Base(root),
			Path: filepath.Clean(root),
		}}
	}

	modulePaths := map[string]bool{}
	for _, module := range modules {
		modulePaths[filepath.Clean(module.Path)] = true
	}
	for i := range modules {
		if err := fillModule(&modules[i], opts.IncludeTests, modulePaths); err != nil {
			return nil, err
		}
		fillModuleBuildMetadata(root, &modules[i], opts.JavaVersion)
		for _, cfg := range modules[i].ConfigFiles {
			props, err := parseConfigFile(cfg)
			if err != nil {
				project.Gaps = append(project.Gaps, model.Gap{
					ID:         stableID("gap", cfg, "config-parse"),
					Message:    "Could not parse config file: " + err.Error(),
					Severity:   "medium",
					Evidence:   evidence(cfg, 1, "config_file", ""),
					Source:     model.SourceFound,
					Confidence: model.ConfidenceMedium,
				})
				continue
			}
			project.ConfigProperties = append(project.ConfigProperties, props...)
		}
	}

	sort.Slice(modules, func(i, j int) bool { return modules[i].Path < modules[j].Path })
	project.Modules = modules
	project.JavaVersion = aggregateJavaVersion(modules)
	project.RefreshSummary()
	return project, nil
}

func discoverModules(root string) ([]model.Module, error) {
	var modules []model.Module
	err := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			if shouldSkipDir(d.Name()) && path != root {
				return filepath.SkipDir
			}
			return nil
		}
		name := d.Name()
		if name != "pom.xml" && name != "build.gradle" && name != "build.gradle.kts" {
			return nil
		}
		dir := filepath.Dir(path)
		modules = append(modules, model.Module{
			ID:        stableID("module", dir),
			Name:      moduleName(root, dir),
			Path:      filepath.Clean(dir),
			BuildTool: buildTool(name),
		})
		return nil
	})
	if err != nil {
		return nil, err
	}
	return dedupeModules(modules), nil
}

func fillModule(module *model.Module, includeTests bool, modulePaths map[string]bool) error {
	sourceRootSet := map[string]bool{}
	err := filepath.WalkDir(module.Path, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			clean := filepath.Clean(path)
			if clean != filepath.Clean(module.Path) && modulePaths[clean] {
				return filepath.SkipDir
			}
			if shouldSkipDir(d.Name()) && path != module.Path {
				return filepath.SkipDir
			}
			return nil
		}
		name := d.Name()
		clean := filepath.Clean(path)
		lower := strings.ToLower(filepath.ToSlash(clean))

		switch {
		case strings.HasSuffix(name, ".java"):
			if !includeTests && isTestPath(lower) {
				return nil
			}
			module.JavaFiles = append(module.JavaFiles, clean)
			if root := sourceRoot(clean); root != "" {
				sourceRootSet[root] = true
			}
		case isConfigFile(name):
			module.ConfigFiles = append(module.ConfigFiles, clean)
		case isMigrationFile(lower):
			module.MigrationFiles = append(module.MigrationFiles, clean)
		case isDescriptorFile(name, lower):
			module.DescriptorFiles = append(module.DescriptorFiles, clean)
		case isUIFile(name):
			module.UIFiles = append(module.UIFiles, clean)
		}
		return nil
	})
	if err != nil {
		return err
	}
	for root := range sourceRootSet {
		module.SourceRoots = append(module.SourceRoots, root)
	}
	sort.Strings(module.SourceRoots)
	sort.Strings(module.JavaFiles)
	sort.Strings(module.ConfigFiles)
	sort.Strings(module.MigrationFiles)
	sort.Strings(module.DescriptorFiles)
	sort.Strings(module.UIFiles)
	return nil
}

func fillModuleBuildMetadata(root string, module *model.Module, overrideJavaVersion string) {
	module.Packaging = inferPackaging(module.Path, module.BuildTool)
	if version := normalizeJavaVersion(overrideJavaVersion); version != "" {
		module.JavaVersion = version
		return
	}
	if version := inferJavaVersionFromAncestors(root, module.Path, module.BuildTool); version != "" {
		module.JavaVersion = version
		return
	}
	module.JavaVersion = "unknown"
}

func inferPackaging(modulePath, buildTool string) string {
	switch buildTool {
	case "maven":
		if packaging := mavenPackaging(filepath.Join(modulePath, "pom.xml")); packaging != "" {
			return packaging
		}
	case "gradle":
		if packaging := gradlePackaging(firstExisting(modulePath, "build.gradle", "build.gradle.kts")); packaging != "" {
			return packaging
		}
	}
	return ""
}

func inferJavaVersionFromAncestors(root, modulePath, buildTool string) string {
	root = filepath.Clean(root)
	for dir := filepath.Clean(modulePath); ; dir = filepath.Dir(dir) {
		if buildTool == "maven" || buildTool == "" {
			if version := mavenJavaVersion(filepath.Join(dir, "pom.xml")); version != "" {
				return version
			}
		}
		if buildTool == "gradle" || buildTool == "" {
			if version := gradleJavaVersion(firstExisting(dir, "build.gradle", "build.gradle.kts")); version != "" {
				return version
			}
		}
		if dir == root {
			break
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
	}
	return ""
}

func aggregateJavaVersion(modules []model.Module) string {
	if len(modules) == 0 {
		return "unknown"
	}
	seen := map[string]bool{}
	for _, module := range modules {
		version := module.JavaVersion
		if version == "" {
			version = "unknown"
		}
		seen[version] = true
	}
	if len(seen) == 1 {
		for version := range seen {
			return version
		}
	}
	return "mixed"
}

func parseConfigFile(path string) ([]model.ConfigProperty, error) {
	ext := strings.ToLower(filepath.Ext(path))
	switch ext {
	case ".properties":
		return parseProperties(path)
	case ".yml", ".yaml":
		return parseYAML(path)
	default:
		return nil, nil
	}
}

func parseProperties(path string) ([]model.ConfigProperty, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var props []model.ConfigProperty
	scanner := bufio.NewScanner(bytes.NewReader(data))
	lineNo := 0
	for scanner.Scan() {
		lineNo++
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") || strings.HasPrefix(line, "!") {
			continue
		}
		idx := strings.IndexAny(line, "=:")
		if idx < 0 {
			continue
		}
		key := strings.TrimSpace(line[:idx])
		value := strings.TrimSpace(line[idx+1:])
		props = append(props, configProperty(path, lineNo, key, value))
	}
	return props, scanner.Err()
}

func parseYAML(path string) ([]model.ConfigProperty, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var root yaml.Node
	if err := yaml.Unmarshal(data, &root); err != nil {
		return nil, err
	}
	var props []model.ConfigProperty
	if len(root.Content) == 0 {
		return props, nil
	}
	flattenYAML(path, root.Content[0], nil, &props)
	return props, nil
}

func flattenYAML(path string, node *yaml.Node, prefix []string, props *[]model.ConfigProperty) {
	switch node.Kind {
	case yaml.MappingNode:
		for i := 0; i+1 < len(node.Content); i += 2 {
			key := node.Content[i]
			value := node.Content[i+1]
			flattenYAML(path, value, append(prefix, key.Value), props)
		}
	case yaml.SequenceNode:
		for i, child := range node.Content {
			flattenYAML(path, child, append(prefix, fmt.Sprintf("%d", i)), props)
		}
	case yaml.ScalarNode:
		key := strings.Join(prefix, ".")
		if key == "" {
			return
		}
		*props = append(*props, configProperty(path, node.Line, key, node.Value))
	}
}

func configProperty(path string, line int, key, value string) model.ConfigProperty {
	definedExternally := strings.Contains(value, "${")
	redacted := isSensitiveKey(key)
	if definedExternally {
		value = "defined_externally"
	} else if redacted {
		value = "redacted"
	}
	return model.ConfigProperty{
		Key:               key,
		Value:             value,
		DefinedExternally: definedExternally,
		Redacted:          redacted,
		SourceFile:        filepath.Clean(path),
		Evidence:          evidence(path, line, "config_property", key),
		Source:            model.SourceFound,
		Confidence:        model.ConfidenceHigh,
	}
}

func isSensitiveKey(key string) bool {
	lower := strings.ToLower(key)
	sensitive := []string{
		"password",
		"passwd",
		"secret",
		"token",
		"access-key",
		"accesskey",
		"private-key",
		"privatekey",
		"credential",
		"credentials",
		"client-secret",
	}
	for _, marker := range sensitive {
		if strings.Contains(lower, marker) {
			return true
		}
	}
	return false
}

func buildTool(name string) string {
	switch name {
	case "pom.xml":
		return "maven"
	case "build.gradle", "build.gradle.kts":
		return "gradle"
	default:
		return ""
	}
}

func moduleName(root, dir string) string {
	rel, err := filepath.Rel(root, dir)
	if err != nil || rel == "." {
		return filepath.Base(root)
	}
	return filepath.ToSlash(rel)
}

func dedupeModules(modules []model.Module) []model.Module {
	seen := map[string]model.Module{}
	for _, module := range modules {
		if existing, ok := seen[module.Path]; ok {
			if existing.BuildTool == "" || module.BuildTool == "maven" {
				existing.BuildTool = module.BuildTool
			}
			seen[module.Path] = existing
			continue
		}
		seen[module.Path] = module
	}
	result := make([]model.Module, 0, len(seen))
	for _, module := range seen {
		result = append(result, module)
	}
	return result
}

func sourceRoot(path string) string {
	slash := filepath.ToSlash(path)
	for _, marker := range []string{"/src/main/java/", "/src/test/java/"} {
		if idx := strings.Index(slash, marker); idx >= 0 {
			return filepath.Clean(slash[:idx+len(marker)-1])
		}
	}
	return ""
}

func shouldSkipDir(name string) bool {
	switch strings.ToLower(name) {
	case ".git", ".gradle", ".idea", ".vscode", "target", "build", "out", "node_modules", "bin", "dist":
		return true
	default:
		return false
	}
}

func isTestPath(path string) bool {
	return strings.Contains(path, "/src/test/") || strings.Contains(path, "/test/")
}

func isConfigFile(name string) bool {
	lower := strings.ToLower(name)
	if lower == "application.properties" || lower == "application.yml" || lower == "application.yaml" {
		return true
	}
	if lower == "bootstrap.properties" || lower == "bootstrap.yml" || lower == "bootstrap.yaml" {
		return true
	}
	return strings.HasPrefix(lower, "application-") && (strings.HasSuffix(lower, ".properties") || strings.HasSuffix(lower, ".yml") || strings.HasSuffix(lower, ".yaml"))
}

func isDescriptorFile(name, path string) bool {
	lowerName := strings.ToLower(name)
	switch lowerName {
	case "web.xml",
		"ejb-jar.xml",
		"application.xml",
		"persistence.xml",
		"beans.xml",
		"webservices.xml",
		"jboss-web.xml",
		"jboss-ejb3.xml",
		"jboss-deployment-structure.xml",
		"jboss-app.xml",
		"ra.xml":
		return true
	}
	return strings.HasSuffix(path, ".wsdl")
}

func isUIFile(name string) bool {
	lower := strings.ToLower(name)
	return strings.HasSuffix(lower, ".xhtml") ||
		strings.HasSuffix(lower, ".jsp") ||
		strings.HasSuffix(lower, ".html") ||
		strings.HasSuffix(lower, ".htm")
}

func isMigrationFile(path string) bool {
	return strings.Contains(path, "/db/migration/") ||
		strings.Contains(path, "/db/changelog/") ||
		strings.Contains(path, "/liquibase/")
}

func firstExisting(dir string, names ...string) string {
	for _, name := range names {
		path := filepath.Join(dir, name)
		if _, err := os.Stat(path); err == nil {
			return path
		}
	}
	return ""
}

func mavenPackaging(path string) string {
	data, ok := readOptionalFile(path)
	if !ok {
		return ""
	}
	return firstXMLTagValue(data, "packaging")
}

func mavenJavaVersion(path string) string {
	data, ok := readOptionalFile(path)
	if !ok {
		return ""
	}
	props := xmlProperties(data)
	for _, tag := range []string{
		"maven.compiler.release",
		"maven.compiler.source",
		"maven.compiler.target",
		"java.version",
		"jdk.version",
		"source",
		"target",
		"release",
	} {
		if version := resolveVersionValue(firstXMLTagValue(data, tag), props); version != "" {
			return version
		}
	}
	return ""
}

func gradlePackaging(path string) string {
	data, ok := readOptionalFile(path)
	if !ok {
		return ""
	}
	lower := strings.ToLower(data)
	switch {
	case strings.Contains(lower, "id 'ear'") || strings.Contains(lower, "id(\"ear\")") || strings.Contains(lower, "apply plugin: 'ear'") || strings.Contains(lower, "apply(plugin = \"ear\")"):
		return "ear"
	case strings.Contains(lower, "id 'war'") || strings.Contains(lower, "id(\"war\")") || strings.Contains(lower, "apply plugin: 'war'") || strings.Contains(lower, "apply(plugin = \"war\")"):
		return "war"
	default:
		return ""
	}
}

func gradleJavaVersion(path string) string {
	data, ok := readOptionalFile(path)
	if !ok {
		return ""
	}
	patterns := []*regexp.Regexp{
		regexp.MustCompile(`(?m)\boptions\.release\s*=\s*([A-Za-z0-9_.'"()]+)`),
		regexp.MustCompile(`(?m)\bsourceCompatibility\s*=\s*([A-Za-z0-9_.'"()]+)`),
		regexp.MustCompile(`(?m)\btargetCompatibility\s*=\s*([A-Za-z0-9_.'"()]+)`),
		regexp.MustCompile(`(?m)\bsourceCompatibility\s+([A-Za-z0-9_.'"()]+)`),
		regexp.MustCompile(`(?m)\btargetCompatibility\s+([A-Za-z0-9_.'"()]+)`),
		regexp.MustCompile(`(?m)\blanguageVersion\s*=\s*JavaLanguageVersion\.of\(([^)]+)\)`),
	}
	for _, pattern := range patterns {
		if match := pattern.FindStringSubmatch(data); len(match) == 2 {
			if version := normalizeJavaVersion(match[1]); version != "" {
				return version
			}
		}
	}
	return ""
}

func readOptionalFile(path string) (string, bool) {
	if path == "" {
		return "", false
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return "", false
	}
	return string(data), true
}

func firstXMLTagValue(data, tag string) string {
	pattern := regexp.MustCompile(`(?is)<` + regexp.QuoteMeta(tag) + `>\s*([^<]+?)\s*</` + regexp.QuoteMeta(tag) + `>`)
	match := pattern.FindStringSubmatch(data)
	if len(match) != 2 {
		return ""
	}
	return strings.TrimSpace(match[1])
}

func xmlProperties(data string) map[string]string {
	props := map[string]string{}
	pattern := regexp.MustCompile(`(?is)<([A-Za-z0-9_.-]+)>\s*([^<]+?)\s*</[A-Za-z0-9_.-]+>`)
	for _, match := range pattern.FindAllStringSubmatch(data, -1) {
		if len(match) != 3 {
			continue
		}
		props[match[1]] = strings.TrimSpace(match[2])
	}
	return props
}

func resolveVersionValue(value string, props map[string]string) string {
	value = strings.TrimSpace(value)
	if strings.HasPrefix(value, "${") && strings.HasSuffix(value, "}") {
		key := strings.TrimSuffix(strings.TrimPrefix(value, "${"), "}")
		value = props[key]
	}
	return normalizeJavaVersion(value)
}

func normalizeJavaVersion(value string) string {
	value = strings.TrimSpace(value)
	value = strings.Trim(value, `"'`)
	value = strings.TrimPrefix(value, "JavaVersion.")
	value = strings.TrimPrefix(value, "VERSION_")
	value = strings.TrimPrefix(value, "VERSION.")
	value = strings.ReplaceAll(value, "_", ".")
	value = strings.TrimSpace(value)
	if value == "" {
		return ""
	}
	if strings.HasPrefix(value, "1.") {
		value = strings.TrimPrefix(value, "1.")
	}
	digits := regexp.MustCompile(`[0-9]+`).FindString(value)
	if digits == "" {
		return ""
	}
	return digits
}

func stableID(parts ...string) string {
	joined := strings.Join(parts, ":")
	replacer := strings.NewReplacer("\\", "/", " ", "-", ":", "-", ".", "-", "_", "-")
	return strings.Trim(replacer.Replace(strings.ToLower(joined)), "-")
}

func evidence(path string, line int, kind, symbol string) model.Evidence {
	return model.Evidence{
		Path:   filepath.Clean(path),
		Line:   line,
		Symbol: symbol,
		Kind:   kind,
	}
}

File: java_process_mapper/discovery.py
from __future__ import annotations

import re
import xml.etree.ElementTree as ET
from pathlib import Path
from typing import Any

from .model import refresh_summary
from .utils import JAVA_VERSION_UNKNOWN, normalize_java_version, read_text, rel_path, stable_id, unique_by_id

SKIP_DIRS = {
    ".git",
    ".gradle",
    ".idea",
    ".mvn",
    ".vscode",
    "__pycache__",
    "bin",
    "build",
    "dist",
    "node_modules",
    "out",
    "target",
}

DESCRIPTOR_NAMES = {
    "application.xml",
    "beans.xml",
    "ejb-jar.xml",
    "faces-config.xml",
    "jboss-deployment-structure.xml",
    "jboss-ejb3.xml",
    "jboss-web.xml",
    "persistence.xml",
    "web-fragment.xml",
    "web.xml",
    "webservices.xml",
}

UI_SUFFIXES = {".xhtml", ".jsp", ".html", ".htm"}


def discover_project(
    root: str | Path,
    *,
    addons: list[str],
    java_version: str | None = None,
    include_tests: bool = False,
) -> dict[str, Any]:
    root_path = Path(root).resolve()
    if not root_path.exists() or not root_path.is_dir():
        raise ValueError(f"root path not found or not a directory: {root_path}")

    module_roots = discover_module_roots(root_path)
    modules = [
        discover_module(root_path, module_root, module_roots, java_version=java_version, include_tests=include_tests)
        for module_root in module_roots
    ]
    source_files = []
    for module in modules:
        source_files.extend(module["sourceFiles"])

    module_versions = {module.get("javaVersion", JAVA_VERSION_UNKNOWN) for module in modules}
    if java_version:
        project_java_version = normalize_java_version(java_version)
    elif len(module_versions) == 1:
        project_java_version = next(iter(module_versions))
    elif module_versions:
        project_java_version = "mixed"
    else:
        project_java_version = JAVA_VERSION_UNKNOWN

    project = {
        "id": stable_id("project", root_path),
        "name": root_path.name,
        "rootPath": str(root_path),
        "addons": addons,
        "javaVersion": project_java_version,
        "modules": modules,
        "sourceFiles": source_files,
        "types": [],
        "entryPoints": [],
        "dependencies": [],
        "graph": {"nodes": [], "edges": []},
    }
    refresh_summary(project)
    return project


def discover_module_roots(root: Path) -> list[Path]:
    candidates: list[Path] = []
    for path in walk_files(root):
        if path.name == "pom.xml" or path.name in {"build.gradle", "build.gradle.kts"}:
            candidates.append(path.parent.resolve())
    if not candidates:
        return [root.resolve()]
    candidates = sorted(set(candidates), key=lambda item: (len(item.parts), str(item).lower()))
    selected: list[Path] = []
    for candidate in candidates:
        if any(candidate == parent for parent in selected):
            continue
        selected.append(candidate)
    return selected


def discover_module(root: Path, module_root: Path, all_module_roots: list[Path], *, java_version: str | None, include_tests: bool) -> dict[str, Any]:
    build_file = find_build_file(module_root)
    build_tool = build_tool_for(build_file)
    build_text = read_text(build_file) if build_file else ""
    packaging = infer_packaging(build_file, build_text, module_root)
    inferred_java = normalize_java_version(java_version) if java_version else infer_java_version(module_root, root)

    module_id = stable_id("module", rel_path(module_root, root) or module_root.name)
    source_files: list[dict[str, Any]] = []
    config_files: list[str] = []
    descriptor_files: list[str] = []
    ui_files: list[str] = []
    migrations: list[str] = []

    for path in walk_files(module_root):
        if belongs_to_nested_module(path, module_root, all_module_roots):
            continue
        if not include_tests and is_test_path(path, module_root):
            continue
        relative = rel_path(path, root)
        suffix = path.suffix.lower()
        if suffix == ".java":
            source_files.append({"path": str(path), "relativePath": relative, "moduleId": module_id, "moduleName": module_root.name})
        elif suffix in {".properties", ".yml", ".yaml", ".xml", ".wsdl"}:
            config_files.append(str(path))
            if path.name in DESCRIPTOR_NAMES or suffix == ".wsdl":
                descriptor_files.append(str(path))
            if is_migration(path):
                migrations.append(str(path))
        elif suffix in UI_SUFFIXES:
            ui_files.append(str(path))

    return {
        "id": module_id,
        "name": module_root.name,
        "rootPath": str(module_root),
        "relativePath": rel_path(module_root, root),
        "buildTool": build_tool,
        "buildFile": str(build_file) if build_file else "",
        "packaging": packaging,
        "javaVersion": inferred_java,
        "sourceFiles": sorted(source_files, key=lambda item: item["path"].lower()),
        "configFiles": sorted(config_files, key=str.lower),
        "descriptorFiles": sorted(set(descriptor_files), key=str.lower),
        "uiFiles": sorted(set(ui_files), key=str.lower),
        "migrations": sorted(set(migrations), key=str.lower),
    }


def belongs_to_nested_module(path: Path, module_root: Path, all_module_roots: list[Path]) -> bool:
    resolved = path.resolve()
    for other in all_module_roots:
        other = other.resolve()
        if other == module_root.resolve():
            continue
        try:
            other.relative_to(module_root.resolve())
            resolved.relative_to(other)
            return True
        except ValueError:
            continue
    return False


def walk_files(root: Path) -> list[Path]:
    result: list[Path] = []
    stack = [root]
    while stack:
        current = stack.pop()
        try:
            children = sorted(current.iterdir(), key=lambda item: item.name.lower())
        except OSError:
            continue
        for child in children:
            if child.is_dir():
                if child.name in SKIP_DIRS:
                    continue
                stack.append(child)
            elif child.is_file():
                result.append(child)
    return result


def is_test_path(path: Path, module_root: Path) -> bool:
    relative = rel_path(path, module_root).lower()
    wrapped = f"/{relative}/"
    return (
        relative.startswith("src/test/")
        or "/src/test/" in wrapped
        or "/test/" in wrapped
        or "/tests/" in wrapped
    )


def find_build_file(module_root: Path) -> Path | None:
    for name in ("pom.xml", "build.gradle", "build.gradle.kts"):
        path = module_root / name
        if path.exists():
            return path
    return None


def build_tool_for(path: Path | None) -> str:
    if path is None:
        return "unknown"
    if path.name == "pom.xml":
        return "maven"
    if path.name.startswith("build.gradle"):
        return "gradle"
    return "unknown"


def infer_packaging(build_file: Path | None, build_text: str, module_root: Path) -> str:
    if build_file and build_file.name == "pom.xml":
        try:
            root = ET.fromstring(build_text)
            packaging = find_xml_text(root, "packaging")
            if packaging:
                return packaging
        except ET.ParseError:
            match = re.search(r"<packaging>\s*([^<]+)\s*</packaging>", build_text)
            if match:
                return match.group(1).strip()
        return "jar"
    if build_file and build_file.name.startswith("build.gradle"):
        lowered = build_text.lower()
        if re.search(r"\bwar\b", lowered):
            return "war"
        if re.search(r"\bear\b", lowered):
            return "ear"
        return "jar"
    if (module_root / "src" / "main" / "webapp").exists():
        return "war"
    return "unknown"


def infer_java_version(module_root: Path, repo_root: Path) -> str:
    current = module_root
    while True:
        for name in ("pom.xml", "build.gradle", "build.gradle.kts"):
            candidate = current / name
            if candidate.exists():
                version = infer_java_version_from_build(candidate)
                if version != JAVA_VERSION_UNKNOWN:
                    return version
        if current == repo_root or current.parent == current:
            break
        current = current.parent
    return JAVA_VERSION_UNKNOWN


def infer_java_version_from_build(path: Path) -> str:
    text = read_text(path)
    if path.name == "pom.xml":
        for pattern in (
            r"<maven\.compiler\.release>\s*([^<]+)\s*</maven\.compiler\.release>",
            r"<maven\.compiler\.source>\s*([^<]+)\s*</maven\.compiler\.source>",
            r"<java\.version>\s*([^<]+)\s*</java\.version>",
            r"<source>\s*([^<]+)\s*</source>",
            r"<release>\s*([^<]+)\s*</release>",
        ):
            match = re.search(pattern, text)
            if match:
                return normalize_java_version(match.group(1))
    for pattern in (
        r"sourceCompatibility\s*=\s*['\"]?([^'\"\s]+)",
        r"targetCompatibility\s*=\s*['\"]?([^'\"\s]+)",
        r"JavaVersion\.VERSION_([A-Za-z0-9_]+)",
        r"languageVersion\s*=\s*JavaLanguageVersion\.of\((\d+)\)",
    ):
        match = re.search(pattern, text)
        if match:
            return normalize_java_version(match.group(1))
    return JAVA_VERSION_UNKNOWN


def find_xml_text(root: ET.Element, local_name: str) -> str:
    for element in root.iter():
        if element.tag.split("}")[-1] == local_name and element.text:
            return element.text.strip()
    return ""


def is_migration(path: Path) -> bool:
    lowered = path.as_posix().lower()
    name = path.name.lower()
    return "db/migration" in lowered or (name.startswith(("v", "r")) and "__" in name and path.suffix.lower() in {".sql", ".xml"})


def module_for_path(project: dict[str, Any], path: str | Path) -> dict[str, Any]:
    resolved = Path(path).resolve()
    modules = sorted(project.get("modules", []), key=lambda item: len(str(item.get("rootPath", ""))), reverse=True)
    for module in modules:
        module_root = Path(module["rootPath"]).resolve()
        try:
            resolved.relative_to(module_root)
            return module
        except ValueError:
            continue
    return project.get("modules", [{}])[0]


def dedupe_project(project: dict[str, Any]) -> None:
    project["entryPoints"] = sorted(
        unique_by_id(project.get("entryPoints", [])),
        key=lambda item: (item.get("framework", ""), item.get("kind", ""), item.get("name", "")),
    )
    project["dependencies"] = sorted(
        unique_by_id(project.get("dependencies", [])),
        key=lambda item: (item.get("kind", ""), item.get("name", "")),
    )
